proposals/NNNN-importing-forward-declared-objc-interfaces-and-protocols.md

```diff
@@ -0,0 +1,177 @@
+such as declaring a new Swift class that inherits from or conforms to such a type. You will also not be able to create new instances of these types in Swift.
+
+To make these limitations more intuitive to users, a new diagnostic has been added. This diagnostic helps guide users when they attempt to access a member on the incomplete synthesized
+type, expecting the complete type. The new diagnostics are shown in the last 4 lines of this example:
+
+```swift
+swift-client-fwd-declared.swift:6:7: error: value of type 'Foo' has no member 'sayHello'
+myFoo.sayHello()
+~~~~~ ^~~~~~~~
+<unknown>:0: note: class 'Foo' is a placeholder for a forward declared Objective-C interface and may be missing members; import the definition to access the complete interface
+foo-bar-consumer.h:3:1: note: interface 'Foo' forward declared here
+@class Foo;
+^
+```
+
+The feature is gated behind a new frontend flag `-enable-import-objc-forward-declarations`. This flag is on by default for Swift version 6 and onwards. The flag is always disabled in the REPL, as
+allowing it leads to confusing behavior. The [existing ClangImporter diagnostics](https://forums.swift.org/t/pitch-improved-clangimporter-diagnostics/52687) will still
+properly warn users about forward declaration limitations, with or with out the feature enabled.
+
+## Source compatibility
+
+This change does have the potential to break source compatibility. This is simply a consequence
+of the ClangImporter importing new declarations that were previously hidden. That said,
+the impact should rare and easy to fix (just rename or qualify any ambiguous identifiers). 
+
+As mentioned, to address this, the feature is gated behind a flag for Swift 5 and before, and enabled
+by default in Swift 6.
+
+That said, the current implementation already passes the source compatibility suite and the source kit stress test.
+
+## Effect on ABI stability
+
+This change has no affect on ABI.
+
+## Effect on API resilience
+
+This change does not influence the ABI characteristics
+of public APIs.
+
+## Alternatives considered
+
+The unavailable attribute attached to synthesized declarations might be more restrictive than necessary.
+We considered the introduction of a new attribute to denote an incomplete type. This attribute would be used to prevent
+operations that are not suitable for an incomplete type (ie. declaring a Swift type that conforms to an incomplete protocol).
+
+Allowing more use of the type however increases the complexity of the problem, and the likelihood of source compatibility issues.
+For example, if the synthesized type is not file private, issues could arise with different files having different representations
+of the type depending on which modules they import.
+
+We also attempted to make this implementation usable from the REPL, but similar issues with conflicting "import environments" arose.
```

Comment:
Once you elaborate on the specifics of what problems you found in the REPL, I think you can just express this as

```suggestion
We could also make this implementation usable from the REPL, but it would require solving the issues described above.
```

And then add a sentence rationalizing why you don't believe solving those should block this proposal.